src/main/scala/org/constellation/gossip/bisect/package.scala

```diff
@@ -1,22 +1,26 @@
 package org.constellation.gossip
 
 import cats.Monad
+import cats.data.OptionT
 import cats.implicits._
-
 package object bisect {
 
-  def bisectA[F[_]: Monad, A](predicate: A => F[Boolean])(seq: IndexedSeq[A]): F[Option[A]] = {
-    val half = Math.floor(seq.length / 2).toInt
-    val (left, right) = seq.splitAt(half)
-
-    predicate(left.last)
-      .ifA(left.pure[F], right.pure[F])
-      .flatMap { partition =>
-        if (partition.size == 1) {
-          partition.headOption.filterA(predicate)
-        } else {
-          bisectA(predicate)(partition)
+  def bisectA[F[_]: Monad, A](predicate: A => F[Boolean], initSeq: Seq[A]): F[Option[A]] = {
+    def go(seq: Seq[A], acc: OptionT[F, A]): OptionT[F, A] = seq match {
```

Comment:
is it possible to make it tail-recursive?

Reply:
Probably yes, but it might be non-trivial. There should be no risk of stack-overflow though, because it's _O(log(n))_

Reply:
if you say so ;) I see a recursion, I think - potential stack overflow :)